File: app/src/main/java/com/diagramqa/app/util/PreferencesManager.kt
package com.diagramqa.app.util

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.runBlocking

private val Context.dataStore by preferencesDataStore(name = "diagramqa_prefs")

/**
 * Persistent app preferences backed by Jetpack DataStore.
 *
 * Stored values:
 *  - dark_mode: 0 = follow system, 1 = on, 2 = off
 *  - animations_enabled: Boolean
 *  - haptics_enabled: Boolean
 */
class PreferencesManager(private val context: Context) {

    suspend fun setDarkMode(mode: Int) {
        context.dataStore.edit { it[DARK_MODE] = mode }
    }

    val darkMode: Int
        get() = runBlocking {
            context.dataStore.data.map { it[DARK_MODE] ?: DARK_MODE_SYSTEM }.first()
        }

    suspend fun setAnimationsEnabled(enabled: Boolean) {
        context.dataStore.edit { it[ANIMATIONS] = enabled }
    }

    val animationsEnabled: Boolean
        get() = runBlocking {
            context.dataStore.data.map { it[ANIMATIONS] ?: true }.first()
        }

    suspend fun setHapticsEnabled(enabled: Boolean) {
        context.dataStore.edit { it[HAPTICS] = enabled }
    }

    val hapticsEnabled: Boolean
        get() = runBlocking {
            context.dataStore.data.map { it[HAPTICS] ?: true }.first()
        }

    companion object {
        const val DARK_MODE_SYSTEM = 0
        const val DARK_MODE_ON = 1
        const val DARK_MODE_OFF = 2

        private val DARK_MODE = intPreferencesKey("dark_mode")
        private val ANIMATIONS = booleanPreferencesKey("animations_enabled")
        private val HAPTICS = booleanPreferencesKey("haptics_enabled")
    }
}

File: src/utils/PreferencesManager.ts
/**
 * Persistent app preferences backed by AsyncStorage.
 * Mirrors PreferencesManager.kt exactly:
 *   dark_mode: 0 = follow system, 1 = on, 2 = off
 *   animations_enabled: boolean (default true)
 *   haptics_enabled: boolean (default true)
 */
import AsyncStorage from '@react-native-async-storage/async-storage';

export const DARK_MODE_SYSTEM = 0;
export const DARK_MODE_ON = 1;
export const DARK_MODE_OFF = 2;

const KEYS = {
  DARK_MODE: 'dark_mode',
  ANIMATIONS: 'animations_enabled',
  HAPTICS: 'haptics_enabled',
};

export const PreferencesManager = {
  async getDarkMode(): Promise<number> {
    const v = await AsyncStorage.getItem(KEYS.DARK_MODE);
    return v !== null ? parseInt(v, 10) : DARK_MODE_SYSTEM;
  },
  async setDarkMode(mode: number): Promise<void> {
    await AsyncStorage.setItem(KEYS.DARK_MODE, String(mode));
  },

  async getAnimationsEnabled(): Promise<boolean> {
    const v = await AsyncStorage.getItem(KEYS.ANIMATIONS);
    return v !== null ? v === 'true' : true;
  },
  async setAnimationsEnabled(enabled: boolean): Promise<void> {
    await AsyncStorage.setItem(KEYS.ANIMATIONS, String(enabled));
  },

  async getHapticsEnabled(): Promise<boolean> {
    const v = await AsyncStorage.getItem(KEYS.HAPTICS);
    return v !== null ? v === 'true' : true;
  },
  async setHapticsEnabled(enabled: boolean): Promise<void> {
    await AsyncStorage.setItem(KEYS.HAPTICS, String(enabled));
  },
};
